apps/chat_oauth2/config/prod.exs

```diff
@@ -14,9 +14,11 @@ use Mix.Config
 # manifest is generated by the mix phx.digest task
 # which you typically run after static files are built.
 config :chat_oauth2, ChatOAuth2Web.Endpoint,
-  load_from_system_env: true,
+  load_from_system_env: false,
```

Comment:
Great point, glad you raised it.

The problem with this setting is its scope. The `load_from_system_env` config option is Phoenix specific. It does allow the phoenix application to read from the system env at runtime.

However, it doesn't make every configuration value readable. For example, any configuration settings for other libraries (like `oauth2_facebook`) are still hardcoded in the config.

Moreover, some of the applications in the umbrella don't use Phoenix at all. Meaning none of their behavior is dynamic.

With this setting turned on, the application behavior is harder to predict. A limited set of config values are read from the env. Which ones again? I know some values are baked in, do I still need to source `.envrc` or not? Etc.

By turning this off, the entire application now acts uniformly. All config options are contained in the single `.envrc` file. All changes take effect the same way.

Finally, the Ansible code that deploys the application uses a templating / placeholder strategy. So `.envrc` values can be still be dynamically updated without rebuilding the release. The only difference is instead of reading them at runtime, they update a templated `sys.config` file.

That way no matter how the application is started everything is loaded as expected. There's no chance of forgetting to source a `.envrc` file. And no confusion about what's dynamic and what isn't.

And agreed on the importance of documenting the deploy / release behaviour thoroughly.

Reply:
Interesting, I think what you are mentioning here is a behavior I haven't seen yet and didn't consider. 

To check my understanding here you are saying that all of our config files for all of our apps, such as `chat_oauth2/config/prod.exs` will honor their `System.get_env` settings BUT libraries such as `oauth2_facebook` in our mix files `chat_oauth2/mix.exs` may or may not honor that and just keep the vars at compile time vs run time no matter which way we set this?

If so that seems really unfortunate from an application developer perspective.

Reply:
The simplest way to include system environment variables is to directly call them in a config file:

```elixir
port: System.get_env("PORT")
```

The downside of this solution is when building a release, the system environment variables get read at _build time_ and are written to the `sys.config` file.

To get around this, a common idiom in Elixir is a two part solution. First, in the config file use a tuple instead:

```elixir
port: {:system, "PORT"}
```

The second piece is in the application itself. Code is written so that whenever it encounters a config value matching `{:system, key}` make a call to `System.get_env(key)`.

The benefit of this solution is at build time the tuple value gets saved into `sys.config`. Then at the application converts the value at runtime.

Hurray, we have dynamic configs. Problem solved! 🎉 

There's a problem though. The two-part solution is by _convention_ only. Every application has to follow the same format rules and implement similar logic for it to work.

In practice, there are many packages / libraries that do not follow this convention. They instead read the value out of the config file. For instance, `oauth2_facebook` falls into this category.

We could update all our apps in the umbrella to follow this pattern. We can even change some of the libraries to support the convention. However, at some point, we're going to hit a wall where something doesn't follow this pattern and certain values aren't updatable at runtime and require rebuilding the release.

That's why the ansible code uses a different solution. It sets `PLACEHOLDER_*` values in `sys.config`. Then when the release is deployed, the placeholder values get replaced with config values.

If values change, they can be updated by running a script locally. No need to rebuild the release. It's not as dynamic as reading `env`s, but it's close. And it makes the config easy to reason about, since everything follows the same pattern.